File: go/src/SimpleTokenBuilder/SimpleTokenBuilder.go
package SimpleTokenBuilder

import (
	"../AccessToken"
	"fmt"
	"time"
	"math/rand"
)

type Role uint16
const (
	Role_Attendee = 1
    Role_Publisher = 2
    Role_Subscriber = 3
    Role_Admin = 4
)

var attendeePrivileges = map[uint16]uint32 {
	AccessToken.KJoinChannel: 0, 
	AccessToken.KPublishAudioStream: 0, 
	AccessToken.KPublishVideoStream: 0, 
	AccessToken.KPublishDataStream: 0,
}
var publisherPrivileges = map[uint16]uint32 {
	AccessToken.KJoinChannel: 0, 
	AccessToken.KPublishAudioStream: 0, 
	AccessToken.KPublishVideoStream: 0, 
	AccessToken.KPublishDataStream: 0,
	AccessToken.KPublishAudiocdn: 0,
	AccessToken.KPublishVideoCdn: 0,
	AccessToken.KInvitePublishAudioStream: 0,
	AccessToken.KInvitePublishVideoStream: 0,
	AccessToken.KInvitePublishDataStream: 0,
}

var subscriberPrivileges = map[uint16]uint32 {
	AccessToken.KJoinChannel: 0, 
	AccessToken.KRequestPublishAudioStream: 0, 
	AccessToken.KRequestPublishVideoStream: 0, 
	AccessToken.KRequestPublishDataStream: 0,
}

var adminPrivileges = map[uint16]uint32 {
	AccessToken.KJoinChannel: 0, 
	AccessToken.KPublishAudioStream: 0, 
	AccessToken.KPublishVideoStream: 0, 
	AccessToken.KPublishDataStream: 0,
	AccessToken.KAdministrateChannel: 0,
}

var RolePrivileges = map[uint16](map[uint16]uint32) {
	Role_Attendee: attendeePrivileges, 
	Role_Publisher: publisherPrivileges, 
	Role_Subscriber: subscriberPrivileges, 
	Role_Admin: adminPrivileges,
}


type SimpleTokenBuilder struct {
	Token AccessToken.AccessToken
}

func random(min int, max int) int {
	rand.Seed(time.Now().UnixNano())
    return rand.Intn(max-min) + min
}

func CreateSimpleTokenBuilder(appID, appCertificate, channelName string, uid uint32) SimpleTokenBuilder {
	uidStr := fmt.Sprintf("%d", uid)
	ts := uint32(time.Now().Unix()) + 24 * 3600
	salt := uint32(random(1, 99999999))
	message := make(map[uint16]uint32)
	token := AccessToken.AccessToken{appID, appCertificate, channelName, uidStr, ts, salt, message}
    return SimpleTokenBuilder{token}
}

func InitPriviliges(builder *SimpleTokenBuilder, role Role) {
	rolepri := uint16(role)
	builder.Token.Message = make(map[uint16]uint32)
	for key, value := range RolePrivileges[rolepri] {
		builder.Token.Message[key] = value
	}
}

func (builder SimpleTokenBuilder) SetPrivilege(privilege AccessToken.Privileges, timeoutFromNow uint32) {
	pri := uint16(privilege)
	builder.Token.Message[pri] = timeoutFromNow
}

func (builder SimpleTokenBuilder) RemovePrivilege(privilege AccessToken.Privileges) {
	pri := uint16(privilege)
	delete(builder.Token.Message, pri)
}

func (builder SimpleTokenBuilder) BuildToken() (string,error) {
	return builder.Token.Build()
}

File: php/src/SimpleTokenBuilder.php
<?php

require "AccessToken.php";

$Role = array(
    "kRoleAttendee" => 0,  // for communication
    "kRolePublisher" => 1, // for live broadcast
    "kRoleSubscriber" => 2,  // for live broadcast
    "kRoleAdmin" => 101
);

$attendeePrivileges = array(
    $Privileges["kJoinChannel"] => 0,
    $Privileges["kPublishAudioStream"] => 0,
    $Privileges["kPublishVideoStream"] => 0,
    $Privileges["kPublishDataStream"] => 0
);


$publisherPrivileges = array(
    $Privileges["kJoinChannel"] => 0,
    $Privileges["kPublishAudioStream"] => 0,
    $Privileges["kPublishVideoStream"] => 0,
    $Privileges["kPublishDataStream"] => 0,
    $Privileges["kPublishAudiocdn"] => 0,
    $Privileges["kPublishVideoCdn"] => 0,
    $Privileges["kInvitePublishAudioStream"] => 0,
    $Privileges["kInvitePublishVideoStream"] => 0,
    $Privileges["kInvitePublishDataStream"] => 0
);

$subscriberPrivileges = array(
    $Privileges["kJoinChannel"] => 0,
    $Privileges["kRequestPublishAudioStream"] => 0,
    $Privileges["kRequestPublishVideoStream"] => 0,
    $Privileges["kRequestPublishDataStream"] => 0
);

$adminPrivileges = array(
    $Privileges["kJoinChannel"] => 0,
    $Privileges["kPublishAudioStream"] => 0,
    $Privileges["kPublishVideoStream"] => 0,
    $Privileges["kPublishDataStream"] => 0,
    $Privileges["kAdministrateChannel"] => 0
);

$RolePrivileges = array(
	$Role["kRoleAttendee"] => $attendeePrivileges,
    $Role["kRolePublisher"] => $publisherPrivileges,
    $Role["kRoleSubscriber"] => $subscriberPrivileges,
    $Role["kRoleAdmin"] => $adminPrivileges
);



class SimpleTokenBuilder
{
    private $token;
    public function __construct($appID, $appCertificate, $channelName, $uid){
        $this->token = new AccessToken($appID, $appCertificate, $channelName, $uid);
    }
    public function initPriviledge($role){
        $p = $RolePrivileges[$role];
        foreach($p as $key => $value){
            $this->setPriviledge($key, $value);
        }
    }
    public function setPriviledge($privilege, $ts){
        $this->token->setPriviledge($privilege, $ts);
    }
    public function removePriviledge($privilege){
        $this->token->removePriviledge($privilege);
    }
    public function buildToken(){
        return $this->token->build();
    }
}


?>